pkg/cloud/metadata.go

```diff
@@ -75,15 +75,20 @@ func (m *Metadata) GetOutpostArn() arn.ARN {
 func NewMetadata() (MetadataService, error) {
 	sess := session.Must(session.NewSession(&aws.Config{}))
 	svc := ec2metadata.New(sess)
-	// creates the in-cluster config
-	config, err := rest.InClusterConfig()
-	if err != nil && !svc.Available() {
-		return nil, err
-	}
-	// creates the clientset
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil && !svc.Available() {
-		return nil, err
+	var clientset *kubernetes.Clientset
+	if !svc.Available() {
+		// creates the in-cluster config
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+		// creates the clientset
+		clientset, err = kubernetes.NewForConfig(config)
+		if err != nil {
+			return nil, err
```

Comment:
let's just omit the test. in runtime, we will either pass NewMetadataService a function svc or a function clientset. There is no situation where we pass neither. So I don't want to muddy the code just for the sake of passing a test case. I will refactor in a future PR so that it's easier to understand , and write another test, before we release this change.

Reply:
what I'm thinking is, NewMetadataService should take an interface with one function. and we will have two implementations of that interface, one based on ec2 metadata and one based on kubernetes API. If ec2 metadata is unavailable you pass the kubernetes API interface. simple!

Reply:
I'm actually not sure how to implement what you mean, I'm not super familiar with Go. So, you might want to make your own PR for that approach. I was originally making this PR because I wanted to highlight that the issue existed, but did not previously see the actual issue reported as a bug.

Reply:
yes @chrisayoub I don't want you to work on more than you signed up for, I can do it in a follow up. sorry for the confusion, I'm thinking out loud.

let's merge your PR without the unit test changes (so just the first commit). 

I can take it from there!
- e2e test it
- refactor
- unit test it

Reply:
do a rebase or whatever to drop every commit but the first, force push it, and I'll merge. The coverage check doesn't matter, I can merge it regardless.

Reply:
Okay does this look okay now?

Reply:
does the fix even work, though? i thought from the unit test attempt we learned that the nil clientset is being cast to a non-nil interface, and therefore bypassing the `if clientset == nil` check altogether.

Reply:
Ah good point. Do you have a suggested fix we can do for this then, or should it just wait for the follow-up refactor?

Reply:
if it works, i'm totally cool with merging without unit tests

Reply:
I'll test it out in our cluster, I put a hold on this until I do that

Reply:
i can't really think of anything besides refactoring the function altogether

Reply:
sounds good. thanks for your help!

Reply:
Done, this should be good to go!